examples/namespace_classes.cf

```diff
@@ -1,32 +1,79 @@
-
-body common control
+bundle agent __main__
 {
-      bundlesequence => { "xxx:globalclass", "main" };
-      inputs => { "namespace_class1.cf" };
+  methods:
+      "MyNamespace:my_bundle";
+
+  reports:
+
+    a_bundle_scoped_class_in_my_namespaced_bundle::
+      "In $(this.namespace):$(this.bundle) I see 'a_bundle_scoped_class_in_my_namespaced_bundle::'";
+
+    !a_bundle_scoped_class_in_my_namespaced_bundle::
+      "In $(this.namespace):$(this.bundle) I do not see 'a_bundle_scoped_class_in_my_namespaced_bundle::'";
+
+    a_namespace_scoped_class_in_my_namespaced_bundle::
+      "In $(this.namespace):$(this.bundle) I see 'a_namespace_scoped_class_in_my_namespaced_bundle::'";
+
+    !a_namespace_scoped_class_in_my_namespaced_bundle::
+      "In $(this.namespace):$(this.bundle) I do not see 'a_namespace_scoped_class_in_my_namespaced_bundle::'";
+
+    MyNamespace:a_bundle_scoped_class_in_my_namespaced_bundle::
+      "In $(this.namespace):$(this.bundle) I see 'MyNamespace:a_bundle_scoped_class_in_my_namespaced_bundle::'";
+
+    !MyNamespace:a_bundle_scoped_class_in_my_namespaced_bundle::
+      "In $(this.namespace):$(this.bundle) I do not see 'MyNamespace:a_bundle_scoped_class_in_my_namespaced_bundle::'";
+
+    MyNamespace:a_namespace_scoped_class_in_my_namespaced_bundle::
+      "In $(this.namespace):$(this.bundle) I see 'MyNamespace:a_namespace_scoped_class_in_my_namespaced_bundle::'";
 }
 
+body file control
+{
+      namespace => "MyNamespace";
+}
 
-bundle agent main
+bundle agent my_bundle
 {
   classes:
+      "a_bundle_scoped_class_in_my_namespaced_bundle";
 
-    cfengine_3::
-
-      "local_default" expression => "any";
+      "a_namespace_scoped_class_in_my_namespaced_bundle"
+        scope => "namespace";
 
-  methods:
+  reports:
 
-      "namespace demo" usebundle => xxx:localclass;
+    a_bundle_scoped_class_in_my_namespaced_bundle::
+      "In $(this.namespace):$(this.bundle) I see 'a_bundle_scoped_class_in_my_namespaced_bundle::'";
 
-  reports:
+    !a_bundle_scoped_class_in_my_namespaced_bundle::
```

Comment:
Interesting you copy/paste the above to here.

What if you had a bundle define before the namespace `body file control` and called it twice? Once from `__main__` and once from another bundle inside the namespace? Maybe too complicated in another way. :/

Reply:
Right, I copy pasta because they are all self referential. If I called one bundle from multiple places it's still one bundle that's called.

But macros for that kind of thing could be a nice feature.

Reply:
@craigcomstock were you thinking of something like this?

    bundle agent __main__
    {
      methods:
          "MyNamespace:my_bundle";
          "report" usebundle => report( "(default)$(this.namespace)", "(main)$(this.bundle)", "called with inherit => 'false'" ), inherit => "false";
     }
     bundle agent report(namespace, bundlename, comment)
     {
      reports:
        a_bundle_scoped_class_in_my_namespaced_bundle::
          "In $(this.namespace):$(this.bundle) From $(namespace):$(bundlename) $(comment) I see 'a_bundle_scoped_class_in_my_namespaced_bundle::'";
        !a_bundle_scoped_class_in_my_namespaced_bundle::
          "In $(this.namespace):$(this.bundle) From $(namespace):$(bundlename) $(comment) I do not see 'a_bundle_scoped_class_in_my_namespaced_bundle::'";
        a_namespace_scoped_class_in_my_namespaced_bundle::
          "In $(this.namespace):$(this.bundle) From $(namespace):$(bundlename) $(comment) I see 'a_namespace_scoped_class_in_my_namespaced_bundle::'";
        !a_namespace_scoped_class_in_my_namespaced_bundle::
          "In $(this.namespace):$(this.bundle) From $(namespace):$(bundlename) $(comment) I do not see 'a_namespace_scoped_class_in_my_namespaced_bundle::'";
        MyNamespace:a_bundle_scoped_class_in_my_namespaced_bundle::
          "In $(this.namespace):$(this.bundle) From $(namespace):$(bundlename) $(comment) I see 'MyNamespace:a_bundle_scoped_class_in_my_namespaced_bundle::'";
        !MyNamespace:a_bundle_scoped_class_in_my_namespaced_bundle::
          "In $(this.namespace):$(this.bundle) From $(namespace):$(bundlename) $(comment) I do not see 'MyNamespace:a_bundle_scoped_class_in_my_namespaced_bundle::'";
        MyNamespace:a_namespace_scoped_class_in_my_namespaced_bundle::
          "In $(this.namespace):$(this.bundle) From $(namespace):$(bundlename) $(comment) I see 'MyNamespace:a_namespace_scoped_class_in_my_namespaced_bundle::'";
    }
    body file control
    {
          namespace => "MyNamespace";
    }
    bundle agent my_bundle
    {
      classes:
          "a_bundle_scoped_class_in_my_namespaced_bundle";
          "a_namespace_scoped_class_in_my_namespaced_bundle"
            scope => "namespace";
         methods:
          "report" usebundle => default:report( "(MyNamespace)$(this.namespace)", "(my_bundle)$(this.bundle)", "called with inherit => 'false'" ), inherit => "false";
          "report" usebundle => default:report( "(MyNamespace)$(this.namespace)", "(my_bundle)$(this.bundle)", "called with inherit => 'true'" ), inherit => "true";
    }

Outputs:

    R: In default:report From default:my_bundle called with inherit => 'false' I do not see 'a_bundle_scoped_class_in_my_namespaced_bundle::'
    R: In default:report From default:my_bundle called with inherit => 'false' I do not see 'a_namespace_scoped_class_in_my_namespaced_bundle::'
    R: In default:report From default:my_bundle called with inherit => 'false' I do not see 'MyNamespace:a_bundle_scoped_class_in_my_namespaced_bundle::'
    R: In default:report From default:my_bundle called with inherit => 'false' I see 'MyNamespace:a_namespace_scoped_class_in_my_namespaced_bundle::'
    R: In default:report From default:my_bundle called with inherit => 'true' I see 'a_bundle_scoped_class_in_my_namespaced_bundle::'
    R: In default:report From default:my_bundle called with inherit => 'true' I do not see 'a_namespace_scoped_class_in_my_namespaced_bundle::'
    R: In default:report From default:my_bundle called with inherit => 'true' I see 'MyNamespace:a_bundle_scoped_class_in_my_namespaced_bundle::'
    R: In default:report From default:my_bundle called with inherit => 'true' I see 'MyNamespace:a_namespace_scoped_class_in_my_namespaced_bundle::'
    R: In default:report From default:main called with inherit => 'false' I do not see 'a_bundle_scoped_class_in_my_namespaced_bundle::'
    R: In default:report From default:main called with inherit => 'false' I do not see 'a_namespace_scoped_class_in_my_namespaced_bundle::'
    R: In default:report From default:main called with inherit => 'false' I do not see 'MyNamespace:a_bundle_scoped_class_in_my_namespaced_bundle::'
    R: In default:report From default:main called with inherit => 'false' I see 'MyNamespace:a_namespace_scoped_class_in_my_namespaced_bundle::'